packages/zoe/src/contractSupport/zoeHelpers.js

```diff
@@ -3,438 +3,363 @@
+  );
+};
 
-        if (!offerSafeForLeft) {
-          console.log(`offer not safe for left`);
-        }
-        if (!offerSafeForRight) {
-          console.log(`offer not safe for right`);
-        }
-        return rejectOffer(tryRight.offerHandle);
-      }
-      zcf.reallocate(
-        [keepLeft.offerHandle, tryRight.offerHandle],
-        [leftAllocation, rightAllocation],
-      );
-      return undefined;
+/**
+ * If the two handles can trade, then swap their compatible assets,
+ * marking both offers as complete.
+ *
+ * The surplus remains with the original offer. For example if
+ * offer A gives 5 moola and offer B only wants 3 moola, offer A
+ * retains 2 moola.
+ *
+ * If the keep offer is no longer active (it was already completed), the try
+ * offer will be rejected with a message (provided by 'keepHandleInactiveMsg').
+ *
+ * TODO: If the try offer is no longer active, swap() should terminate with
+ * a useful error message, like defaultRejectMsg.
+ *
+ * If the swap fails, no assets are transferred, and the 'try' offer is rejected.
+ *
+ * @param {ZCFSeat} keepSeat
+ * @param {ZCFSeat} trySeat
+ * @param {String} [keepHandleInactiveMsg]
+ */
+export const swap = (
+  zcf,
+  keepSeat,
+  trySeat,
+  keepHandleInactiveMsg = 'prior offer is unavailable',
+) => {
+  if (keepSeat.hasExited()) {
+    throw trySeat.kickOut(keepHandleInactiveMsg);
+  }
+
+  trade(
+    zcf,
+    {
+      seat: keepSeat,
+      gains: keepSeat.getProposal().want,
+    },
+    {
+      seat: trySeat,
+      gains: trySeat.getProposal().want,
     },
+  );
 
-    /**
-     * If the two handles can trade, then swap their compatible assets,
-     * marking both offers as complete.
-     *
-     * The surplus remains with the original offer. For example if
-     * offer A gives 5 moola and offer B only wants 3 moola, offer A
-     * retains 2 moola.
-     *
-     * If the keep offer is no longer active (it was already completed), the try
-     * offer will be rejected with a message (provided by 'keepHandleInactiveMsg').
-     *
-     * TODO: If the try offer is no longer active, swap() should terminate with
-     * a useful error message, like defaultRejectMsg.
-     *
-     * If the swap fails, no assets are transferred, and the 'try' offer is rejected.
-     *
-     * @param {OfferHandle} keepHandle
-     * @param {OfferHandle} tryHandle
-     * @param {String} [keepHandleInactiveMsg]
-     */
-    swap: (
-      keepHandle,
-      tryHandle,
-      keepHandleInactiveMsg = 'prior offer is unavailable',
-    ) => {
-      if (!zcf.isOfferActive(keepHandle)) {
-        throw helpers.rejectOffer(tryHandle, keepHandleInactiveMsg);
+  keepSeat.exit();
+  trySeat.exit();
+  return defaultAcceptanceMsg;
+};
+
+/**
+ * @typedef ExpectedRecord
+ * @property {Record<keyof ProposalRecord['give'],null>} [want]
+ * @property {Record<keyof ProposalRecord['want'],null>} [give]
+ * @property {Partial<Record<keyof ProposalRecord['exit'],null>>} [exit]
+ */
```

Comment:
> I don't understand why some of these are partial and other not, even though all of them are optional.

`exit` is partial because not every kind of exit needs appear.  For example both `waived` and `onDemand` would otherwise be required, which causes a type failure if you didn't supply one of them:

```
src/contracts/auction/secondPriceAuction.js:87:7 - error TS2739: Type '{ waived: null; }' is missing the following properties from type 'Record<"waived" | "onDemand" | "afterDeadline", null>': onDemand, afterDeadline

87       exit: { waived: null },
         ~~~~

  src/contractSupport/zoeHelpers.js:219:4
    219  * @property {Record<keyof ProposalRecord['exit'], null>} [exit]
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    The expected type comes from property 'exit' which is declared here on type 'ExpectedRecord'
```

> Is `keyof ProposalRecord['give']` the same thing as `Keyword`? If so, `Keyword` would be clearer.

Done.